app/models/piece.rb

```diff
@@ -15,22 +15,68 @@ def serializable_hash(options = nil)
     super.merge("type" => type)
   end
 
-  def move_to!(row, col)
-    raise "Out of bounds" if row < 0 || row > 7 || col < 0 || col > 7
-    create_move!(row, col)
-    update(row: row, col: col)
-  end
+  def move_to!(to_row, to_col)
+    # rubocop: disable LineLength
+    Rails.logger.debug "Model move_to. from: #{row},#{col} to: #{to_row},#{to_col}"
+
+    if valid_move?(to_row, to_col)
+      from_row = row   # needed for create_move if valid_move? true
+      from_col = col   # needed for create_move if valid_move? true
+      Rails.logger.debug "valid_move result TRUE. Piece stays at new location."
+      update_attributes(row: to_row, col: to_col)
+      create_move!(from_row, from_col)
+    else
+      Rails.logger.debug "valid move result FALSE. Piece pop back when games show page reloaded."
+    end # method move_to!
+  end   # method move_to!
 
   private
 
-  def create_move!(new_row, new_col)
+  def valid_move?(to_row, to_col)
+    return false if move_nil?(to_row, to_col)
+    return false if move_out_of_bounds?(to_row, to_col)
+    return false unless move_legal?(to_row, to_col)
```

Comment:
Actually, castling may have to be implemented in a separate method that gets called before `move_obstructed?`.